### ISSUE TITLE
A constraint cannot involve a massless particle

### ISSUE THREAD

**Issue by des2037** [openmm/openmm#2783]
Hi everyone,

I'm trying to 'freeze' some atoms of a protein's backbone at their starting coordinates by setting the atoms' masses to 0.0000. However, I'm hitting an exception:
```
Exception: A constraint cannot involve a massless particle
```

Given this message, I'm assuming the error is due to the use of this constraint:
```
constraints = AllBonds
```
(this is done to enable a 4 fs timestep)

Is there a way to constrain the coordinates of an atom, when all bonds are constrained? If not, any suggestions for a possible workaround?  

Thanks very much,
Derek

**Comment by peastman** [comment]
You'll need to remove any constraints that involve a massless particle.  You can do something like this.

```python
for i in reversed(range(system.getNumConstraints())):
  p1, p2, d = system.getConstraintParameters(i)
  if system.getParticleMass(p1) == 0*amu or system.getParticleMass(p2) == 0*amu:
    system.removeConstraint(i)

**Comment by des2037** [comment]
Thanks very much @peastman !!! Not only does that totally resolve my issue (without having to resort to super-massive atoms + harmonic constraints), but these `system` functions are new to me, which is exciting. : -)

Just one caveat that might be useful for anyone who encounters this issue: when I constrained all of the backbone atoms for a given residue (including the backbone hydrogens), the system blows up:
`Exception: Particle coordinate is nan` 

This surprised me (not sure why massless hydrogens were a problem), but practically working around this is easy. Specifically, I didn't run into any problem when I just made the backbone atoms that aren't connected to any hydrogens massless (e.g. the carbonyl carbon, carbonyl oxygen, amine nitrogen for prolines, etc.).

Thanks again!
Derek
